packages/account-sdk/src/interface/public-utilities/prolink/utils/encoding.ts

```diff
@@ -0,0 +1,217 @@
+  }
+
+  let hex = '';
+  for (let i = 0; i < bytes.length; i++) {
+    hex += bytes[i].toString(16).padStart(2, '0');
+  }
+
+  return BigInt(`0x${hex}`);
+}
+
+/**
+ * Encode capabilities map to protobuf format
+ * @param caps - Capabilities object
+ * @returns Map with UTF-8 JSON-encoded values
+ */
+export function encodeCapabilities(caps: Record<string, unknown>): Map<string, Uint8Array> {
+  const map = new Map<string, Uint8Array>();
+
+  for (const [key, value] of Object.entries(caps)) {
+    const json = JSON.stringify(value);
+    const bytes = new TextEncoder().encode(json);
+    map.set(key, bytes);
+  }
+
+  return map;
+}
+
+/**
+ * Decode capabilities map from protobuf format
+ * @param map - Map with UTF-8 JSON-encoded values
+ * @returns Capabilities object
+ */
+export function decodeCapabilities(map: Map<string, Uint8Array>): Record<string, unknown> {
+  const caps: Record<string, unknown> = {};
+
+  for (const [key, bytes] of map.entries()) {
+    try {
+      const json = new TextDecoder().decode(bytes);
+      caps[key] = JSON.parse(json);
+    } catch (error) {
+      throw new Error(
+        `Failed to decode capability '${key}': ${error instanceof Error ? error.message : 'unknown error'}`
+      );
+    }
+  }
+
+  return caps;
+}
+
+/**
+ * Pad a value to 32 bytes (for EIP-712 encoding)
+ * @param bytes - Value to pad
+ * @returns 32-byte padded value
+ */
+export function pad32(bytes: Uint8Array): Uint8Array {
+  if (bytes.length > 32) {
+    throw new Error(`Cannot pad value larger than 32 bytes: ${bytes.length}`);
+  }
+
+  const padded = new Uint8Array(32);
+  // Left-pad with zeros
+  padded.set(bytes, 32 - bytes.length);
+  return padded;
+}
+
+/**
+ * Convert bytes to hex string with 0x prefix
+ * Minimal encoding: no leading zeros unless value is zero
+ * @param bytes - Bytes to convert
+ * @returns Hex string
+ */
+export function bytesToHex(bytes: Uint8Array): string {
+  if (bytes.length === 0) {
+    return '0x0';
+  }
+
+  // For single byte 0, return 0x0
+  if (bytes.length === 1 && bytes[0] === 0) {
+    return '0x0';
+  }
+
+  let hex = '0x';
+  let foundNonZero = false;
+
+  for (let i = 0; i < bytes.length; i++) {
+    // Skip leading zero bytes (but keep the last byte even if it's zero)
+    if (!foundNonZero && bytes[i] === 0 && i < bytes.length - 1) {
+      continue;
+    }
+    foundNonZero = true;
+
+    // For the first non-zero byte, don't pad if it's a single digit
+    if (foundNonZero && hex === '0x' && bytes[i] < 16) {
+      hex += bytes[i].toString(16);
+    } else {
+      hex += bytes[i].toString(16).padStart(2, '0');
+    }
+  }
+
+  return hex;
+}
+
+/**
+ * Convert hex string to bytes
+ * @param hex - Hex string (with or without 0x prefix)
+ * @returns Bytes
+ */
+export function hexToBytes(hex: string): Uint8Array {
+  const normalized = hex.toLowerCase().replace(/^0x/, '');
+  const bytes = new Uint8Array(normalized.length / 2);
```

Comment:
If hex is `"0x123"`, then `normalized.length = 3`, and `bytes = new Uint8Array(1.5)` becomes `new Uint8Array(1)`, so we will lose data here.

I would recommend using `Bytes`, `Base64`, etc from [Ox](https://oxlib.sh) for all these utilities fwiw since they are battle-tested and more performant.

Reply:
You're right, odd length hex is broken here. Adding test cases for coverage then applying a fix.